### PR TITLE
Low-vision accessibility

### PR DESCRIPTION
**Pull request by imobachgs**
This PR should fix [bsc#768112](https://bugzilla.novell.com/show_bug.cgi?id=768112). 

## Features

* Colors are changed using style sheets instead of modifying the palette (when the old code was written, QSS didn't exist). This approach will allow to modify not only colors but also font sizes, spacing, etc.
* Switching to high-contrast mode during installation (pressing `shift+F4`) works again. A default `highcontrast.qss` style sheet should be available.
* A new `Y2HIGHCONTRAST` variable have been introduced:
  * This variable will be used to boot the installer in the high-contrast mode from the beginning.
  * You can set the name of a style sheet to be used as the high-contrast alternative. That would allow us to ship different style sheets to be used by people with different low-vision troubles (for example, using different colors, sizes, etc.).
  * If you set the variable to a non-existent QSS file (like `Y2HIGHCONTRAST=1` or `Y2HIGHCONTRAST=somewrongvalue) the default `highcontrast.qss` file will be used.
* If no high-contrast QSS file can be found, nothing will happen. Previously, due to [bsc#768112](https://bugzilla.novell.com/show_bug.cgi?id=768112), a pop-up was shown telling the user that colors were changed to high-contrast mode although they weren't changed. It also means that we can merge this code and shift+f4 will look as disabled until we provide a `highcontrast.qss` stylesheet.

## Screenshot

**The stylesheet is not ready yet, but we're working on it.**

high-contrast-qss

## Why an additional variable?

This code makes `libyui-qt` aware of the `Y2HIGHCONTRAST` environment variable. In the beginning, I was reluctant to make `libyui-qt` aware of that variable and I used `Y2STYLE`, but it had some limitations:

* The only way to detect that the installer was started in high-contrast mode was to compare `Y2STYLE` with the hard-coded value `highcontrast.qss`. So to ship different high-contrast stylesheets, we would need to hard-code all the alternatives.
* Given that the `Y2STYLE` was set to `highcontrast.qss`, shif+f4 should be disabled in this mode (because we didn't know the name of the default stylesheet).

### CODE REVIEW

**Comment by kobliha** [comment]
Very nice :) :+1:

**Comment by lslezak** [comment]
An idea: Linuxrc (or kernel?) sets the environment from the boot command line. If you boot with `FOO=bar` boot option then you have `$FOO` with `bar` value in the installer.

We could use `Y2STYLE=highcontrast.qss` boot option to start the installation with an alternative style right from the beginning.

Unfortunately the starting script hardcodes (overrides) the `Y2STYLE` value: https://github.com/yast/yast-yast2/blob/master/scripts/yast2-funcs#L133. But the fix should be easy, use `installation.qss` only if `Y2STYLE` is not already set...

What do you think about that?

**Comment by imobachgs** [comment]
Yes, you're right, that's something we definitively should do (similar to [Gabi](https://bugzilla.novell.com/show_bug.cgi?id=780621#c7) suggestion). I added the `currentStyleSheet` just for that reason: to detect the current stylesheet just in case it was overriden from the beginning using the `Y2STYLE` variable (and we're in *high-contrast* mode from the beginning). And as you have said, it should be easy to adapt the `Y2STYLE` behavior.

On the other hand, we should integrate both interfaces: ncurses and Qt interfaces. Gabi reported that for ncurses we need to set the `Y2NCURSES_COLOR_THEME` variable to `highcontrast`. So we could take care of setting this variable if `Y2STYLE` is enabled (or some similar behaviour).

Finally, maybe we could add a new option to Linuxrc (as F7 or whatever) to set the installer in highcontrast mode. That would be easier than typing the variable.

Opinions?

**Comment by lslezak** [comment]
Yes, unification with the ncurses UI would be nice. The question is how to implement that.

- Add a new variable (`Y2HIGHCONTRAST=1` or something like that)
- Set `Y2NCURSES_COLOR_THEME=highcontrast` when `Y2STYLE` is equal to `highcontrast.qss` and the text mode is used

I'm not sure which one is better. The first one adds a new option and we already have quite a lot of them. The second is not that intuitive and a bit weird, we would need to document that well.

I do not know if a new `F7` boot menu item would be nice, there are already six menus. That's quite a lot and this one will not be used often. What about adding it to the `F4 Video Mode` menu? That makes sense to me, you can choose the video mode (resolution, ...) and you could switch to the high contrast as well.

But the new `F7` menu would be really prominent and a good selling point, so from marketing POV it would make sense...

@wfeldt how difficult is to tweak the installation menu as described above? What would be better? The `F7` or the `F4` menu?

**Comment by shundhammer** [comment]
+1 for adding it to the F4 Video Mode menu to make it well accessible.

**Comment by imobachgs** [comment]
About the unification, maybe I would go for the `Y2HIGHCONTRAST=1` option (I agree that it's more intuitive).

Regarding the menu:

* Current options in F4 look like exclusive to me (`Text Mode` or `No KMS` or `Default`). What if you want to use `Text Mode` + `High Contrast`? Should we have a separate option for that?
* Having in a different `F7` key makes it easier to reach for the user. But I can see you point that maybe there already too much menus.

**Comment by lslezak** [comment]
Yes, on the first sight the `F4` menu looks like a radio button where only one can be active. But it could be implement similarly to the `F2` where you choose the language (definitely only one can be active).  But there is an additional `Keyboard` submenu where you can select a different keyboard layout for the language.

We could have a `Color` or `UI Style` submenu with `Default` and `High Contrast` values.

@wfeldt ?

**Comment by kobliha** [comment]
Just a few notes from PO POV:

- New variable is OK, e.g., Y2HIGHCONTRAST
- Making decision on the fact whether something equals to `highcontrast.qss` (or similar condition) is a way to break it in the future, it should be simple
- Qt and ncurses should both use the same variable, meaning that Linuxrc should support the same variable for both (unified easy to remember, harder to break)
- Adding a new entry to Linuxrc menu is OK, but another possibility can be considere too (Linuxrc screen support was requested), but better consult this with Ken first
- Linuxrc menu refactoring is also possible

**Comment by lslezak** [comment]
*[Just for the record, I have asked Ken via email to comment here.]*

**Comment by wfeldt** [comment]
If you change the gfxboot menu, I'd go for a submenu in F4. More so as the video menu will need a refresh anyway. F7 is not a good idea as - you might not have seen it :-)  - there already is an F7-menu to switch architectures on bi-arch DVDs.

Independent of this we will need a comparatively simple command line option to pass the setting on to linuxrc/yast. On EFI systems there is no gfxboot.

**Comment by wfeldt** [comment]
Just another note: we might then as well put the screen reader functionality that's activated by F9 into an F4 submenu, so people actually know about it...

**Comment by lslezak** [comment]
Oh, sorry, actually the video mode menu in SP2 is at the `F3` key, not `F4`. (I do not know whether I overlooked that or it's HW dependent or I booted some old version by mistake...).

Anyway, it should be in the `Video Mode` menu.

Steffen, thanks for describing the F7 and F9 keys, I didn't know that...

File: package/libyui-qt.changes [hunk 1]
```diff
@@ -1,4 +1,10 @@
 -------------------------------------------------------------------
+Mon Oct 10 16:56:41 UTC 2016 - igonzalezsosa@suse.com
+
+- Fix high-contrast support (bsc#76811 and related tobsc#780621)
+- 2.46.26
```
Comment (shundhammer) [inline package/libyui-qt.changes]: .spec says 2.46.27

File: src/QY2Styler.cc [hunk 2]
```diff
@@ -64,39 +70,73 @@ QY2Styler::styler()
...
 
-	if ( ! style.isEmpty() )
-	    styler->loadStyleSheet( style );
-	else
-	    styler->loadStyleSheet( "style.qss" );
+void QY2Styler::setDefaultStyleSheet(const QString & styleSheet)
+{
+    QFileInfo fileInfo(themeDir() + styleSheet);
+
+    if (fileInfo.isFile()) {
```
Comment (shundhammer) [inline src/QY2Styler.cc]: wrong indentation style for the UI - it always puts the { on the next line

File: src/QY2Styler.cc [hunk 3]
```diff
@@ -64,39 +70,73 @@ QY2Styler::styler()
...
+                   << _defaultStyleSheet << std::endl;
     }
+}
 
-    return styler;
+void QY2Styler::setHighContrastStyleSheet(const QString & styleSheet)
+{
+    QFileInfo fileInfo(themeDir() + styleSheet);
+
+    if (fileInfo.isFile()) {
```
Comment (shundhammer) [inline src/QY2Styler.cc]: Again: wrong indentation style - { should be on the next line

File: package/libyui-qt.changes [hunk 4]
```diff
@@ -1,4 +1,10 @@
 -------------------------------------------------------------------
+Mon Oct 10 16:56:41 UTC 2016 - igonzalezsosa@suse.com
+
+- Fix high-contrast support (bsc#76811 and related tobsc#780621)
```
Comment (shundhammer) [inline package/libyui-qt.changes]: and one blank missing: "to bsc#"

**Comment by shundhammer** [comment]
LGTM

**Comment by mvidner** [comment]
@imobachgs it seems that there is still some code using the API that was removed:
```console
[   70s] /home/abuild/rpmbuild/BUILD/libyui-qt-pkg-2.45.7/src/YQPkgGenericDetailsView.cc: In static member function 'static QString YQPkgGenericDetailsView::htmlHeading(ZyppSel, bool)':
[   70s] /home/abuild/rpmbuild/BUILD/libyui-qt-pkg-2.45.7/src/YQPkgGenericDetailsView.cc:168:24: error: 'class YQUI' has no member named 'usingVisionImpairedPalette'
[   70s]      if ( ! YQUI::ui()->usingVisionImpairedPalette() )
[   70s]                         ^
[   70s] /home/abuild/rpmbuild/BUILD/libyui-qt-pkg-2.45.7/src/YQPkgGenericDetailsView.cc: In static member function 'static QString YQPkgGenericDetailsView::table(const QString&)':
[   70s] /home/abuild/rpmbuild/BUILD/libyui-qt-pkg-2.45.7/src/YQPkgGenericDetailsView.cc:206:24: error: 'class YQUI' has no member named 'usingVisionImpairedPalette'
[   70s]      if ( ! YQUI::ui()->usingVisionImpairedPalette() )
[   70s]                         ^
```
https://ci.suse.de/view/libyui/job/libyui-qt-pkg-SLE-12/38/console

**Comment by imobachgs** [comment]
@mvidner Oh, I'll check it.

**Pull request merged**